File: typescript/src/modules/log.ts
/**
 * LogModule — Send structured logs from SDK apps to the RealtimeX system log.
 *
 * Logs are sent asynchronously (fire-and-forget) so they never block your app.
 * Failed sends are silently ignored to avoid cascading errors.
 *
 * @example
 * ```ts
 * sdk.log.info("My app started");
 * sdk.log.error("LLM call failed", { error: e.message, model: "gpt-4o" });
 * ```
 */

export type LogLevel = "debug" | "info" | "warn" | "error" | "fatal";
export type LogSource = "app" | "server" | "llm" | "vector-db" | "channel" | "embed" | "external";

export interface LogEntry {
  level: LogLevel;
  source: LogSource;
  message: string;
  context?: Record<string, unknown>;
  timestamp?: string;
}

export interface LogResponse {
  success: boolean;
  ingested: number;
}

export class LogModule {
  private readonly baseUrl: string;
  private readonly appId: string;
  private readonly apiKey?: string;
  private queue: LogEntry[] = [];
  private flushTimer: ReturnType<typeof setTimeout> | null = null;
  private readonly FLUSH_INTERVAL_MS = 2000;
  private readonly BATCH_SIZE = 20;

  constructor(baseUrl: string, appId: string, apiKey?: string) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.appId = appId;
    this.apiKey = apiKey;
  }

  private get headers(): Record<string, string> {
    const h: Record<string, string> = { "Content-Type": "application/json" };
    if (this.apiKey) h["Authorization"] = `Bearer ${this.apiKey}`;
    if (this.appId) h["x-app-id"] = this.appId;
    return h;
  }

  /** Flush queued logs immediately — returns true on success */
  async flush(): Promise<boolean> {
    if (this.queue.length === 0) return true;
    const batch = this.queue.splice(0, this.BATCH_SIZE);
    try {
      const res = await fetch(`${this.baseUrl}/sdk/logs`, {
        method: "POST",
        headers: this.headers,
        body: JSON.stringify({ logs: batch }),
      });
      return res.ok;
    } catch {
      // Fire-and-forget: silently discard on network error
      return false;
    }
  }

  private schedule(): void {
    if (this.flushTimer !== null) return;
    this.flushTimer = setTimeout(() => {
      this.flushTimer = null;
      this.flush().catch(() => {});
    }, this.FLUSH_INTERVAL_MS);
  }

  /** Queue a log entry and schedule a flush */
  log(level: LogLevel, message: string, context?: Record<string, unknown>, source: LogSource = "app"): void {
    this.queue.push({ level, source, message, context, timestamp: new Date().toISOString() });
    if (this.queue.length >= this.BATCH_SIZE) {
      if (this.flushTimer !== null) { clearTimeout(this.flushTimer); this.flushTimer = null; }
      this.flush().catch(() => {});
    } else {
      this.schedule();
    }
  }

  debug(message: string, context?: Record<string, unknown>, source: LogSource = "app"): void {
    this.log("debug", message, context, source);
  }

  info(message: string, context?: Record<string, unknown>, source: LogSource = "app"): void {
    this.log("info", message, context, source);
  }

  warn(message: string, context?: Record<string, unknown>, source: LogSource = "app"): void {
    this.log("warn", message, context, source);
  }

  error(message: string, context?: Record<string, unknown>, source: LogSource = "app"): void {
    this.log("error", message, context, source);
  }

  fatal(message: string, context?: Record<string, unknown>, source: LogSource = "app"): void {
    this.log("fatal", message, context, source);
  }
}

File: python/realtimex_sdk/log.py
"""
LogModule — Send structured logs from SDK apps to the RealtimeX system log.

Logs are sent asynchronously (fire-and-forget) so they never block your app.
"""
import asyncio
import threading
from typing import Optional, List, Dict, Any, Literal

LogLevel = Literal["debug", "info", "warn", "error", "fatal"]
LogSource = Literal["app", "server", "llm", "vector-db", "channel", "embed", "external"]

FLUSH_INTERVAL = 2.0  # seconds
BATCH_SIZE = 20


class LogModule:
    """
    Sends structured logs from an SDK app to the RealtimeX system log.

    Usage::

        sdk.log.info("App started")
        sdk.log.error("Something failed", {"error": str(e)})
    """

    def __init__(self, base_url: str, app_id: str, api_key: Optional[str] = None):
        self._base_url = base_url.rstrip("/")
        self._app_id = app_id
        self._api_key = api_key
        self._queue: List[Dict[str, Any]] = []
        self._lock = threading.Lock()
        self._timer: Optional[threading.Timer] = None

    @property
    def _headers(self) -> Dict[str, str]:
        h = {"Content-Type": "application/json"}
        if self._api_key:
            h["Authorization"] = f"Bearer {self._api_key}"
        if self._app_id:
            h["x-app-id"] = self._app_id
        return h

    def _flush_sync(self, batch: List[Dict[str, Any]]) -> None:
        try:
            import httpx
            httpx.post(
                f"{self._base_url}/sdk/logs",
                json={"logs": batch},
                headers=self._headers,
                timeout=5.0,
            )
        except Exception:
            pass  # Fire-and-forget

    def _schedule(self) -> None:
        if self._timer is not None:
            return
        self._timer = threading.Timer(FLUSH_INTERVAL, self._on_timer)
        self._timer.daemon = True
        self._timer.start()

    def _on_timer(self) -> None:
        with self._lock:
            self._timer = None
            if not self._queue:
                return
            batch = self._queue[:BATCH_SIZE]
            self._queue = self._queue[BATCH_SIZE:]
        threading.Thread(target=self._flush_sync, args=(batch,), daemon=True).start()

    def log(
        self,
        level: LogLevel,
        message: str,
        context: Optional[Dict[str, Any]] = None,
        source: LogSource = "app",
    ) -> None:
        """Queue a log entry for sending."""
        from datetime import datetime, timezone
        entry = {
            "level": level,
            "source": source,
            "message": str(message)[:5000],
            "context": context,
            "timestamp": datetime.now(timezone.utc).isoformat(),
        }
        with self._lock:
            self._queue.append(entry)
            if len(self._queue) >= BATCH_SIZE:
                batch = self._queue[:BATCH_SIZE]
                self._queue = self._queue[BATCH_SIZE:]
                if self._timer:
                    self._timer.cancel()
                    self._timer = None
                threading.Thread(target=self._flush_sync, args=(batch,), daemon=True).start()
                return
            self._schedule()

    def debug(self, message: str, context: Optional[Dict[str, Any]] = None, source: LogSource = "app") -> None:
        self.log("debug", message, context, source)

    def info(self, message: str, context: Optional[Dict[str, Any]] = None, source: LogSource = "app") -> None:
        self.log("info", message, context, source)

    def warn(self, message: str, context: Optional[Dict[str, Any]] = None, source: LogSource = "app") -> None:
        self.log("warn", message, context, source)

    def error(self, message: str, context: Optional[Dict[str, Any]] = None, source: LogSource = "app") -> None:
        self.log("error", message, context, source)

    def fatal(self, message: str, context: Optional[Dict[str, Any]] = None, source: LogSource = "app") -> None:
        self.log("fatal", message, context, source)

    async def flush(self) -> bool:
        """Async flush — sends all queued logs immediately."""
        with self._lock:
            if not self._queue:
                return True
            batch = self._queue[:BATCH_SIZE]
            self._queue = self._queue[BATCH_SIZE:]
        try:
            import httpx
            async with httpx.AsyncClient() as client:
                res = await client.post(
                    f"{self._base_url}/sdk/logs",
                    json={"logs": batch},
                    headers=self._headers,
                    timeout=5.0,
                )
                return res.is_success
        except Exception:
            return False
